### PR TITLE
Fork resolution when Python requirement is narrowed

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This is similar to #4707, but in this case, we don't actually fork, because there are no divergent requirements. Consider:

```text
uv ; python_version >= "3.8"
```

With `-p 3.7`. In this case, there's no version of `uv` that satisfies Python 3.7, but it's only required on Python 3.8 anyway.

The solution I settled on here is to... fork anyway? I'd like to get some input before setting it out, but it makes some sense to me: we want to solve with a different set of constraints than in the rest of the resolution.

Initially, I tried to just merge the fork markers with the requirement markers (i.e., the `python_version >= "3.8"` on the requirement itself) -- but that's insufficient. Imagine if instead we had a package `foo` that depended on `uv`, and `foo ; python_version >= "3.8"`.

Closes https://github.com/astral-sh/uv/issues/4668.

### CODE REVIEW

**Comment by charliermarsh** [comment]
I'm honestly not sure if this is sound.

File: crates/uv/tests/branching_urls.rs [hunk 1]
```diff
@@ -397,7 +397,6 @@ fn root_package_splits_other_dependencies_too() -> Result<()> {
         { name = "anyio", version = "4.2.0", source = { registry = "https://pypi.org/simple" }, marker = "python_version < '3.12'" },
         { name = "anyio", version = "4.3.0", source = { registry = "https://pypi.org/simple" }, marker = "python_version >= '3.12'" },
         { name = "b1", marker = "python_version < '3.12'" },
-        { name = "b2", marker = "python_version >= '3.12'" },
```
Comment (charliermarsh) [inline crates/uv/tests/branching_urls.rs:400]: Ok yeah this is just wrong.

**Comment by charliermarsh** [comment]
Open to better ideas here. I think what I have in this PR is wrong actually.

**Comment by konstin** [comment]
I think the splitting can work, but we must only split when there the python marker is stricter than the current requires-python and be mindful if we're already splitting on the python version marker, which is happening `root_package_splits_other_dependencies_too`.

**Comment by konstin** [comment]
Is this still applicable with #6143?

**Pull request closed without merging**